source/common/stats/stats_impl.h

```diff
@@ -61,11 +74,58 @@ class TagProducerImpl : public TagProducer {
   std::string produceTags(const std::string& metric_name, std::vector<Tag>& tags) const override;
 
 private:
+  friend class DefaultTagRegexTester;
+
+  /**
+   * Adds a TagExtractor to the collection of tags, tracking prefixes to help make
+   * produceTags run efficiently by trying only extractors that have a chance to match.
+   * @param extractor TagExtractorPtr the extractor to add.
+   */
+  void addExtractor(TagExtractorPtr extractor);
+
+  /**
+   * Adds all default extractors matching the specified tag name. In this model,
+   * more than one TagExtractor can be used to generate a given tag. The default
+   * extractors are specified in common/config/well_known_names.cc.
+   * @param name absl::string_view the extractor to add.
+   */
+  void addExtractorsMatching(absl::string_view name);
+
+  /**
+   * Roughly estimate the size of the vectors.
+   * @param config const envoy::config::metrics::v2::StatsConfig& the config.
+   */
   void reserveResources(const envoy::config::metrics::v2::StatsConfig& config);
-  void addDefaultExtractors(const envoy::config::metrics::v2::StatsConfig& config,
-                            std::unordered_set<std::string>& names);
 
-  std::vector<TagExtractorPtr> tag_extractors_;
+  /**
+   * Adds all default extractors from well_known_names.cc into the
+   * collection. Returns a set of names of all default extractors
+   * into a string-set for dup-detection against new stat names
+   * specified in the configuration.
+   * @param config const envoy::config::metrics::v2::StatsConfig& the config.
+   * @return names std::unordered_set<std::string> the set of names to populate
+   */
+  std::unordered_set<std::string>
+  addDefaultExtractors(const envoy::config::metrics::v2::StatsConfig& config);
+
+  /**
+   * Iterates over every tag extractor that might possibly match stat_name, calling
```

Comment:
nit: "that might possibly match stat_name" seems a little unspecific.  Do you think it might make sense to be specific about what exact criterion this method checks against (prefix for those that have it) to determine the set over which the function is called?

Reply:
I'm coy about this implementation detail in the interface doc because I am weighing whether it makes sense to also scan for substrings.  In particular, the rq_xxx paterns are slow, and can be eliminated with a quick scan for "rq" on the input string, so it may not be prefixes in the future.  I want to do some more targeted tests for that after more of the low hanging fruit is plucked.

Reply:
Hmmm, since the primary goal of the method is to select on certain criteria, I think that should be explicitly laid out in the documentation since the only other way to determine it is to check the implementation.  Could we maybe create a bullet list of the criteria here, and for now it'll just show the prefix, and later we can add to that if things change?  Also, I thought only public methods required doxygen documentation in Envoy.  Has this changed?

Reply:
RE show what actually happens in the current impl, and update when that changes: SGTM. Coming.

RE doxygen: I would like to know what the actual post-processing flow is.  I think there's a bunch of redundancy in the doxygen comments.  In the past when I've used doxygen for C++ I've done it somewhat differently, commenting the vars in the formal-list, so you don't have to repeat the variable name or the type.  So right now I'm just trying to be consistent but I wonder what the overall benefit is.

However @alyssawilk asked specifically for more function-doc here and in general in Envoy and I agree with that.  It helps the reader to more easily follow unfamiliar code so I'm all over that.  I wouldn't mind being somewhat terser though.

Reply:
I think the history around Doxygen usage in Envoy can be summed up as the fact that Doxygen style was adopted as a convention, but rather than actually running Doxygen on the source Envoy development has historically involved just human reading the Doxygen comments in the header files. So the quirks in usage of Doxygen style can generally be traced to this somewhat docs usage pattern.